block-producer/src/batch_builder/mod.rs

```diff
@@ -18,66 +25,105 @@ mod tests;
 ///
 /// Note: Until recursive proofs are available in the Miden VM, we don't include the common proof.
 pub struct TransactionBatch {
-    txs: Vec<SharedProvenTx>,
+    account_initial_states: Vec<(AccountId, Digest)>,
+    updated_accounts: Vec<(AccountId, Digest)>,
+    consumed_notes_script_roots: Vec<Digest>,
+    produced_nullifiers: Vec<Digest>,
+    created_notes_smt: SimpleSmt,
 }
 
 impl TransactionBatch {
-    pub fn new(txs: Vec<SharedProvenTx>) -> Self {
-        Self { txs }
+    pub fn new(txs: Vec<SharedProvenTx>) -> Result<Self, BuildBatchError> {
+        let account_initial_states =
+            txs.iter().map(|tx| (tx.account_id(), tx.initial_account_hash())).collect();
+        let updated_accounts =
+            txs.iter().map(|tx| (tx.account_id(), tx.final_account_hash())).collect();
+
+        let consumed_notes_script_roots = {
+            let mut script_roots: Vec<Digest> = txs
+                .iter()
+                .flat_map(|tx| tx.consumed_notes())
+                .map(|consumed_note| consumed_note.script_root())
+                .collect();
+
+            script_roots.sort();
+
+            // Removes duplicates in consecutive items
+            script_roots.into_iter().dedup().collect()
+        };
+        let produced_nullifiers = txs
+            .iter()
+            .flat_map(|tx| tx.consumed_notes())
+            .map(|consumed_note| consumed_note.nullifier())
+            .collect();
+
+        let created_notes_smt = {
+            let created_notes: Vec<_> = txs
+                .iter()
+                .flat_map(|tx| tx.created_notes())
+                .map(|note_envelope| note_envelope.note_hash())
+                .collect();
+
+            if created_notes.len() > MAX_NUM_CREATED_NOTES_PER_BATCH {
+                return Err(BuildBatchError::TooManyNotesCreated(created_notes.len()));
+            }
+
+            SimpleSmt::with_leaves(
+                CREATED_NOTES_SMT_DEPTH,
+                created_notes.into_iter().enumerate().map(|(idx, note_hash)| {
+                    (
+                        idx.try_into().expect("already checked not for too many notes"),
+                        note_hash.into(),
+                    )
+                }),
+            )?
+        };
```

Comment:
I think this is slightly off. I believe the leaves in created note SMT are computed as `hash(note_hash || note_metadata)`. This also means that the length of the path from not root to note hash is 21.

Also, nit: I'd probably move this logic into a separate helper function.

Reply:
Will adjust. But why aren't we instead using a tree of depth 20, but where the leaf nodes' hash is computed as `hash(note_hash || note_metadata)` (easily computed from a `NoteEnvelope`)?

The current structure is a bit awkward, since
+ the depth of the tree is not linked with the number of notes anymore
  + i.e. `2^21 != max number of notes`
+ leaves always come in pair, as opposed to being independent (as we typically think of leaves)

Reply:
> But why aren't we instead using a tree of depth 20, but where the leaf nodes' hash is computed as `hash(note_hash || note_metadata)` (easily computed from a `NoteEnvelope`)?

Good question. Let's discuss this - maybe indeed using a tree of depth 20 will simplify things.

Reply:
Fixed - the batches now have depth 13, and each `NoteEnvelope` is stored as 2 nodes in the tree.

> Also, nit: I'd probably move this logic into a separate helper function.

I personally disagree with this. IMO, unless a piece of code is used more than once (or is really long/complex), it shouldn't be its own function. Mainly because naming is *hard*; so you're unlikely to find good names for the helper functions, and ultimately as a reader of the code, I need to scroll out of context to go check the details of the function.

I actually find that Rust making code blocks expressions (such that you can define vars as `let v = { ... };` is one of Rust's best "small" features, exactly because you no longer need to define many helper functions - the indentation helps visually separate the logic necessary to define `v`, and temporary variables are not accessible outside the block. This is IMO an improvement on javascript's self-invoking functions, where I would write `var v = (function() { ... })();` to achieve the same result (i.e. avoiding the need to name helper functions, and bring the code out of its context).

Would love your thoughts on this - ideally, we'd adopt a consistent convention across our code bases (i.e. use of blocks vs use of helper functions for cases like this one).

Reply:
Ah - interesting! I haven't thought about it this way before.

I usually try to move code out into dedicated functions when it encompass some non-trivial well-encapsulated logic. I think this improves readability because it clearly defines interfaces and reduces the number of things someone needs to understand when reading code.

In the above example, understanding how notes SMT is built is not material to understanding how a transaction batch is instantiated (we just need to know that we need to build a note SMT) - so, in my mind it made sense to move it to a separate function. On the other hand, the code is simple enough that we can probably keep it as is.

Overall, I try not to have functions which span more than 50 - 60 lines - though, it doesn't apply to this case either as the whole function is just 40.

Reply:
> I think this improves readability because it clearly defines interfaces and reduces the number of things someone needs to understand when reading code.

Right, this is probably the best counterpoint to my argument, and yes that's fair.

> On the other hand, the code is simple enough that we can probably keep it as is.

Right, it's now quite simple. But I agree that if it were a bit more complex, I would move to a helper function, and I'll adopt that convention moving forward.